File: WindowsCE/NurSample_vs2008_VB/NurSampleCE_VB/AntennaTuner.vb
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Drawing
Imports System.Data
Imports System.Text
Imports System.Windows.Forms
Imports System.Runtime.InteropServices

Imports NurApiDotNet
'NurApi wrapper
Partial Public Class AntennaTuner
    Inherits UserControl
    Private hNur As NurApi = Nothing

    Public Sub New()
        InitializeComponent()
    End Sub

    Private Sub NurTune_Load(ByVal sender As Object, ByVal e As EventArgs)

    End Sub

    ''' <summary>
    ''' Sets the NurApi.
    ''' </summary>
    ''' <param name="hNur">The handle of NurApi.</param>
    Public Sub SetNurApi(ByVal hNur As NurApi)
        Try
            Me.hNur = hNur

            ' Set event handlers for NurApi
            AddHandler hNur.DisconnectedEvent, New EventHandler(Of NurApi.NurEventArgs)(AddressOf hNur_DisconnectedEvent)
            AddHandler hNur.ConnectedEvent, New EventHandler(Of NurApi.NurEventArgs)(AddressOf hNur_ConnectedEvent)

            ' Update the status of the connection
            If hNur.IsConnected() Then
                hNur_ConnectedEvent(hNur, Nothing)
            Else
                hNur_DisconnectedEvent(hNur, Nothing)
            End If
        Catch ex As NurApiException
            MessageBox.Show(ex.ToString(), Program.appName)
        End Try
    End Sub

    ''' <summary>
    ''' Handles the DisconnectedEvent event of the NUR module.
    ''' </summary>
    ''' <param name="sender">The source of the event.</param>
    ''' <param name="e">The <see cref="NurApi.NurEventArgs" /> instance containing the event data.</param>
    Private Sub hNur_DisconnectedEvent(ByVal sender As Object, ByVal e As NurApi.NurEventArgs)
        Dim hNur As NurApi = TryCast(sender, NurApi)
        Me.Enabled = False
    End Sub

    ''' <summary>
    ''' Handles the ConnectedEvent event of the NUR module.
    ''' </summary>
    ''' <param name="sender">The source of the event.</param>
    ''' <param name="e">The <see cref="NurApi.NurEventArgs" /> instance containing the event data.</param>
    Private Sub hNur_ConnectedEvent(ByVal sender As Object, ByVal e As NurApi.NurEventArgs)
        Dim hNur As NurApi = TryCast(sender, NurApi)
        Try
            Me.Enabled = NurCapabilities.I.DeviceCaps.tune
            If NurCapabilities.I.DeviceCaps.tune = False Then
                AddMessage("Not available with this module!!!")
            End If
        Catch generatedExceptionName As NurApiException
            Me.Enabled = False
        End Try
    End Sub

    ''' <summary>
    ''' Handles the Click event of the measureReflegtedPowerButton control.
    ''' </summary>
    ''' <param name="sender">The source of the event.</param>
    ''' <param name="e">The <see cref="EventArgs"/> instance containing the event data.</param>
    Private Sub measureReflegtedPowerButton_Click(ByVal sender As Object, ByVal e As EventArgs)
        AddMessage("*** Measure Reflected Powers *** ")
        MeasureReflectedPowers()
    End Sub

    ''' <summary>
    ''' Handles the Click event of the tuneButton control.
    ''' </summary>
    ''' <param name="sender">The source of the event.</param>
    ''' <param name="e">The <see cref="EventArgs"/> instance containing the event data.</param>
    Private Sub tuneButton_Click(ByVal sender As Object, ByVal e As EventArgs)
        AddMessage("*** Measure Reflected Powers *** ")
        MeasureReflectedPowers()
        AddMessage("*** Tune Antennas *** ")
        TuneAntennas()
        AddMessage("*** Measure Reflected Powers *** ")
        MeasureReflectedPowers()
    End Sub

    Private Sub restoreFactoryTuningsButtons_Click(ByVal sender As Object, ByVal e As EventArgs)
        AddMessage("*** Measure Reflected Powers *** ")
        MeasureReflectedPowers()
        AddMessage("*** Restore Factory Tuning *** ")
        RestoreTuning()
        AddMessage("*** Measure Reflected Powers *** ")
        MeasureReflectedPowers()
    End Sub


    ''' <summary>
    ''' Tunes the antennas.
    ''' </summary>
    Private Sub TuneAntennas()
        Try
            Dim antenna As Integer = 0
            Dim mask As Integer = hNur.AntennaMask
            If mask > 0 Then
                While mask > 0
                    If (mask And 1) <> 0 Then
                        For band As Integer = 0 To 5
                            Try
                                Dim duration As Integer = System.Environment.TickCount
                                Dim reply As NurTuner.TuneResult()
                                reply = NurTuner.Tune(hNur, NurTuner.TYPE.WIDE, antenna, band, -50, True)
                                duration = System.Environment.TickCount - duration
                                ShowResults(reply, duration)
                            Catch ex As NurApiException
                                AddMessage("Error " + ex.[error] + ". ")
                            End Try
                            HHUtils.KeepDeviceAlive()
                        Next
                        mask = mask >> 1
                        antenna += 1
                    End If
                End While
                AddMessage("Done")
            End If
        Catch ex As NurApiException
            AddMessage("NurApiException " + ex.[error])
        End Try
    End Sub

    Private Sub ShowResults(ByVal resp As NurTuner.TuneResult(), ByVal duration As Integer)
        Dim sb As New StringBuilder()
        sb.Append(String.Format("Ant{0}, ", resp(0).antenna))
        For i As Integer = 0 To resp.Length - 1
            sb.Append(String.Format("Bnd{0}={1}, ", resp(i).band, resp(i).ToString()))
        Next
        sb.Append(String.Format("{0} ms", duration))
        AddMessage(sb.ToString())
    End Sub

    ''' <summary>
    ''' Measures the Reflected Powers from enabled antennas.
    ''' </summary>
    Private Sub MeasureReflectedPowers()
        Try
            AddMessage(Convert.ToString("Region: ") & hNur.GetRegionInfo().name)

            ' Measure Reflected Power
            Dim antenna As Integer = 0
            Dim mask As Integer = hNur.AntennaMask
            Dim sb As New StringBuilder()
            If mask > 0 Then
                Dim tmpSelectdeAntenna As Integer = hNur.SelectedAntenna
                While mask > 0
                    If (mask And 1) <> 0 Then
                        Try
                            ' Select Antenna
                            hNur.SelectedAntenna = antenna
                            ' Measure Reflected Power
                            'NurApi.RegionInfo regionInfo = hNur.GetRegionInfo();
                            'hNur.GetReflectedPower(regionInfo.baseFrequency); // FW bug fix
                            Dim rpInfo As NurApi.ReflectedPowerInfo = hNur.GetReflectedPower()
                            ' Calculate Reflected Power
                            Dim rf As Double = Math.Sqrt(CDbl(rpInfo.iPart * rpInfo.iPart + rpInfo.qPart * rpInfo.qPart))
                            rf /= CDbl(rpInfo.div)
                            rf = Math.Log10(rf) * 20.0
                            If [Double].IsInfinity(rf) Then
                                rf = -30
                            End If
                            sb.Append(String.Format("Ant{0}: {1:0.0} dBm, ", antenna, rf))
                        Catch generatedExceptionName As NurApiException
                            sb.Append(String.Format("Ant{0}: ?.? dBm, ", antenna))
                        End Try
                    End If
                    mask = mask >> 1
                    antenna += 1
                End While
                ' Restore SelectedAntenna
                hNur.SelectedAntenna = tmpSelectdeAntenna
            Else
                sb.Append("No eabled antennas!")
            End If
            sb.Replace(", ", "", sb.Length - 2, 2)
            AddMessage(sb.ToString())
            AddMessage("Done")
        Catch ex As NurApiException
            AddMessage("NurApiException " + ex.[error])
        End Try
    End Sub

    ''' <summary>
    ''' Tunes the antennas.
    ''' </summary>
    Private Sub RestoreTuning()
        Try
            hNur.RestoreTuning(True)
            AddMessage("Done")
        Catch ex As NurApiException
            AddMessage("NurApiException " + ex.[error])
        End Try
    End Sub

    ''' <summary>
    ''' Adds message to listBox.
    ''' </summary>
    ''' <param name="message">The message.</param>
    Private Sub AddMessage(ByVal message As String)
        listBox1.Items.Add(message)
        listBox1.SelectedIndex = listBox1.Items.Count - 1
        Application.DoEvents()
    End Sub
End Class

File: NurSamplePC_vs2008_CS/NurSamplePC/AntennaTuner.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Text;
using System.Windows.Forms;
using System.Runtime.InteropServices;

using NurApiDotNet; //NurApi wrapper

namespace NurSample
{
    public partial class AntennaTuner : UserControl
    {
        NurApi hNur = null;

        public AntennaTuner()
        {
            InitializeComponent();
        }

        private void NurTune_Load(object sender, EventArgs e)
        {

        }

        /// <summary>
        /// Sets the NurApi.
        /// </summary>
        /// <param name="hNur">The handle of NurApi.</param>
        public void SetNurApi(NurApi hNur)
        {
            try
            {
                this.hNur = hNur;

                // Set event handlers for NurApi
                hNur.DisconnectedEvent += new EventHandler<NurApi.NurEventArgs>(hNur_DisconnectedEvent);
                hNur.ConnectedEvent += new EventHandler<NurApi.NurEventArgs>(hNur_ConnectedEvent);
                hNur.TuneEvent += new EventHandler<NurApi.TuneEventArgs>(hNur_TuneEvent);

                // Update the status of the connection
                if (hNur.IsConnected())
                    hNur_ConnectedEvent(hNur, null);
                else
                    hNur_DisconnectedEvent(hNur, null);
            }
            catch (NurApiException ex)
            {
                MessageBox.Show(ex.ToString(), Program.appName);
            }
        }

        void hNur_TuneEvent(object sender, NurApi.TuneEventArgs e)
        {
            AddMessage(string.Format("TuneEvent Ant{0}: {1} dBm",
                e.data.antenna, (decimal)e.data.reflPower_dBm/1000));
        }

        /// <summary>
        /// Handles the DisconnectedEvent event of the NUR module.
        /// </summary>
        /// <param name="sender">The source of the event.</param>
        /// <param name="e">The <see cref="NurApi.NurEventArgs" /> instance containing the event data.</param>
        private void hNur_DisconnectedEvent(object sender, NurApi.NurEventArgs e)
        {
            NurApi hNur = sender as NurApi;
            this.Enabled = false;
        }

        /// <summary>
        /// Handles the ConnectedEvent event of the NUR module.
        /// </summary>
        /// <param name="sender">The source of the event.</param>
        /// <param name="e">The <see cref="NurApi.NurEventArgs" /> instance containing the event data.</param>
        private void hNur_ConnectedEvent(object sender, NurApi.NurEventArgs e)
        {
            NurApi hNur = sender as NurApi;
            try
            {
                this.Enabled = NurCapabilities.I.DeviceCaps.tune;
                if (NurCapabilities.I.DeviceCaps.tune == false)
                    AddMessage("Not available with this module!!!");
            }
            catch (NurApiException)
            {
                this.Enabled = false;
            }
        }

        /// <summary>
        /// Handles the Click event of the measureReflegtedPowerButton control.
        /// </summary>
        /// <param name="sender">The source of the event.</param>
        /// <param name="e">The <see cref="EventArgs"/> instance containing the event data.</param>
        private void measureReflegtedPowerButton_Click(object sender, EventArgs e)
        {
            AddMessage("*** Measure Reflected Powers *** ");
            MeasureReflectedPowers();
        }

        /// <summary>
        /// Handles the Click event of the tuneButton control.
        /// </summary>
        /// <param name="sender">The source of the event.</param>
        /// <param name="e">The <see cref="EventArgs"/> instance containing the event data.</param>
        private void tuneButton_Click(object sender, EventArgs e)
        {
            AddMessage("*** Measure Reflected Powers *** ");
            MeasureReflectedPowers();
            AddMessage("*** Tune Antennas *** ");
            AddMessage("Please wait...");
            TuneAntennas();
            AddMessage("*** Measure Reflected Powers *** ");
            MeasureReflectedPowers();
        }

        private void restoreFactoryTuningsButtons_Click(object sender, EventArgs e)
        {
            AddMessage("*** Measure Reflected Powers *** ");
            MeasureReflectedPowers();
            AddMessage("*** Restore Factory Tuning *** ");
            RestoreTuning();
            AddMessage("*** Measure Reflected Powers *** ");
            MeasureReflectedPowers();
        }


        /// <summary>
        /// Tunes the antennas.
        /// </summary>
        private void TuneAntennas()
        {
            try
            {
                bool tuneEventsEnabled = hNur.EnableTuneEvents;
                if (tuneEventsEnabled)
                    hNur.EnableTuneEvents = false;
                List<int> enabledAntennas = hNur.EnabledAntennas;
                foreach (int antenna in enabledAntennas)
                {
                    try
                    {
                        int duration = System.Environment.TickCount;
                        int []reply = hNur.TuneAntenna(antenna, true, true);
                        duration = System.Environment.TickCount - duration;
                        ShowResults(antenna, reply, duration);
                    }
                    catch (NurApiException ex)
                    {
                        AddMessage("Error " + ex.error + ". ");
                    }
                }                    
                AddMessage("Done");
                
                if (tuneEventsEnabled)
                    hNur.EnableTuneEvents = true;
            }
            catch (NurApiException ex)
            {
                AddMessage("NurApiException " + ex.error);
            }
        }

        private void ShowResults(int antenna, int[] tuneResp, int duration)
        {
            StringBuilder sb = new StringBuilder();
            sb.Append(string.Format("Ant{0}, ", antenna));
            for (int i = 0; i < tuneResp.Length; i++)
            {
                sb.Append(string.Format("Bnd{0}={1}, ", i, ((float)tuneResp[i] / 1000.0).ToString()));
            }
            sb.Append(string.Format("{0} ms", duration));
            AddMessage(sb.ToString());
        }

        /// <summary>
        /// Measures the Reflected Powers from enabled antennas.
        /// </summary>
        void MeasureReflectedPowers()
        {
            try
            {
                bool tuneEventsEnabled = hNur.EnableTuneEvents;
                if (tuneEventsEnabled)
                    hNur.EnableTuneEvents = false;
                AddMessage("Region: " + hNur.GetRegionInfo().name);

                int tmpSelectdeAntenna = hNur.SelectedAntenna;
                List<int> enabledAntennas = hNur.EnabledAntennas;
                StringBuilder sb = new StringBuilder();
                foreach (int antenna in enabledAntennas)
                {
                    try
                    {
                        // Select Antenna
                        hNur.SelectedAntenna = antenna;
                        // Measure Reflected Power
                        double dBm = hNur.GetReflectedPowerValue(0);
                        sb.Append(string.Format("Ant{0}: {1:0.0} dBm, ", antenna, dBm));
                    }
                    catch (NurApiException)
                    {
                        sb.Append(string.Format("Ant{0}: ?.? dBm, ", antenna));
                    }
                }                
                sb.Replace(", ", "", sb.Length - 2, 2);
                AddMessage(sb.ToString());
                AddMessage("Done");

                // Restore SelectedAntenna
                hNur.SelectedAntenna = tmpSelectdeAntenna;

                if (tuneEventsEnabled)
                    hNur.EnableTuneEvents = true;
            }
            catch (NurApiException ex)
            {
                AddMessage("NurApiException " + ex.error);
            }
        }

        /// <summary>
        /// Tunes the antennas.
        /// </summary>
        private void RestoreTuning()
        {
            try
            {
                hNur.RestoreTuning(true);
                AddMessage("Done");
            }
            catch (NurApiException ex)
            {
                AddMessage("NurApiException " + ex.error);
            }
        }

        /// <summary>
        /// Adds message to listBox.
        /// </summary>
        /// <param name="message">The message.</param>
        void AddMessage(string message)
        {
            listBox1.Items.Add(message);
            listBox1.SelectedIndex = listBox1.Items.Count - 1;
            Application.DoEvents();
        }
    }
}
